modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/SchemaProvider.scala

```diff
@@ -12,48 +12,28 @@ import cats.effect.Sync
 import cats.implicits._
 import com.snowplowanalytics.iglu.client.{ClientError, Resolver}
 import com.snowplowanalytics.iglu.client.resolver.registries.RegistryLookup
-import com.snowplowanalytics.iglu.core.{SchemaCriterion, SchemaKey, SchemaMap, SelfDescribingSchema}
+import com.snowplowanalytics.iglu.client.resolver.Resolver.SchemaResolutionError
+import com.snowplowanalytics.iglu.core.{SchemaKey, SchemaMap, SelfDescribingSchema}
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.Schema
 import com.snowplowanalytics.iglu.schemaddl.jsonschema.circe.implicits.toSchema
 import com.snowplowanalytics.snowplow.badrows.FailureDetails
 
-import scala.math.Ordered._
-
 private[transform] object SchemaProvider {
 
-  private def getSchema[F[_]: Sync: RegistryLookup](
-    resolver: Resolver[F],
-    schemaKey: SchemaKey
-  ): EitherT[F, FailureDetails.LoaderIgluError, SelfDescribingSchema[Schema]] =
-    for {
-      json <- EitherT(resolver.lookupSchema(schemaKey))
-                .leftMap(resolverBadRow(schemaKey))
-      schema <- EitherT.fromOption[F](Schema.parse(json), parseSchemaBadRow(schemaKey))
-    } yield SelfDescribingSchema(SchemaMap(schemaKey), schema)
-
   // Note schema order of the returned list is not guaranteed
   def fetchSchemasWithSameModel[F[_]: Sync: RegistryLookup](
     resolver: Resolver[F],
     schemaKey: SchemaKey
   ): EitherT[F, FailureDetails.LoaderIgluError, NonEmptyVector[SelfDescribingSchema[Schema]]] =
     for {
-      schemaKeys <- EitherT(resolver.listSchemasLike(schemaKey))
-                      .leftMap(resolverFetchBadRow(schemaKey.vendor, schemaKey.name, schemaKey.format, schemaKey.version.model))
-                      .map(_.schemas.toVector)
-      schemaKeys <- EitherT.rightT[F, FailureDetails.LoaderIgluError](schemaKeys.filter(_ < schemaKey))
-      topSchema <- getSchema(resolver, schemaKey)
-      lowerSchemas <- schemaKeys.filter(_ < schemaKey).traverse(getSchema(resolver, _))
-    } yield NonEmptyVector(topSchema, lowerSchemas)
-
-  private def resolverFetchBadRow(
-    vendor: String,
-    name: String,
-    format: String,
-    model: Int
-  )(
-    e: ClientError.ResolutionError
-  ): FailureDetails.LoaderIgluError =
-    FailureDetails.LoaderIgluError.SchemaListNotFound(SchemaCriterion(vendor = vendor, name = name, format = format, model = model), e)
+      jsons <- EitherT(resolver.lookupSchemasUntil(schemaKey))
+                 .leftMap { case SchemaResolutionError(schemaKey, error) => resolverBadRow(schemaKey)(error) }
+      schemas <- jsons.traverse { json =>
+                   EitherT
+                     .fromOption[F](Schema.parse(json.schema), parseSchemaBadRow(json.self.schemaKey))
+                     .map(schema => SelfDescribingSchema(SchemaMap(json.self.schemaKey), schema))
+                 }
+    } yield NonEmptyVector(schemas.head, schemas.tail.toVector)
```

Comment:
This makes me nervous to see an unsafe method like `schemas.head` which raises an exception if `schemas` is empty.

I know I know `schemas` should never be empty, because we know that `resolver.lookupSchemasUntil` should never return an empty list.  But ideally we would make that explicit in types, instead of saying "I the programmer know this is safe".

Can you think of any way to avoid it?  For example, could `lookupSchemasUntil` return a `NonEmpty*` data structure?

Reply:
This is a good point. I'll update `lookupSchemasUntil` to make it return a `NonEmptyVector`, like this function

Reply:
Hey @istreeter , for `lookupSchemasUntil` I wonder if it's better to work directly with `Vector` or to start with a `List` and then convert it. The prepend on the `List` is faster but at the end we need to revert it whereas with `Vector` we can have it directly in the right order. Given that we can expect only a few elements both should be fine but I'm curious if you have an opinion ?

Reply:
I would probably make `lookupSchemasUntil` return a `NonEmptyList` and also make `SchemaProvider.fetchSchemasWithSameModel` return a `NonEmptyList`.  Because, based on a quick look at the implementations, it looks like List is quite a natural return type for how they are implemented.

An alternative implementation is to have a return type of `NonEmptyVector` for both, and use vectors everywhere in the implementation.  That would also be ok.

I try to avoid needlessly converting between list and vector.  However, I care more when it's a method that is invoked once (or more) per event; whereas the methods here are called once per _batch_ of events.  So I am slightly more relaxed about which data structure to pick.

Reply:
> make SchemaProvider.fetchSchemasWithSameModel return a NonEmptyList

It's a breaking change in the API, are we fine with that ? Won't it require to convert from `NonEmptyList` to `NonEmptyVector` in the app that calls this function ?

Reply:
First of all, it is not a breaking change because `SchemaProvider` is private to common-streams.

Furthermore, common-streams is still pre-1.x and we are making breaking changes in every release so far.  I am planning many other breaking changes!

[Here is](https://github.com/snowplow-incubator/common-streams/blob/0.7.0/modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/NonAtomicFields.scala#L67) the only place where we use the result returned from `SchemaProvider.fetchSchemasWithSameModel` and it looks completely ok to use either a list or a vector there.